lib/knapsack_pro/runners/queue/rspec_runner.rb

```diff
@@ -1,226 +1,255 @@
+            def log_rspec_command(args, test_file_paths, type)
+              case type
+              when :subset_queue
+                KnapsackPro.logger.info("To retry the last batch of tests fetched from the API Queue, please run the following command on your machine:")
+              when :end_of_queue
+                KnapsackPro.logger.info("To retry all the tests assigned to this CI node, please run the following command on your machine:")
+              end
 
-              KnapsackPro::Hooks::Queue.call_after_subset_queue
+              stringified_cli_args = args.join(' ')
+              FORMATTERS.each do |formatter|
+                stringified_cli_args.sub!(" --format #{formatter}", '')
+              end
 
-              return {
-                status: :next,
-                runner: runner,
-                can_initialize_queue: false,
-                args: args,
-                exitstatus: exitstatus,
-                all_test_file_paths: all_test_file_paths,
-              }
+              KnapsackPro.logger.info(
+                "bundle exec rspec #{stringified_cli_args} " +
+                KnapsackPro::TestFilePresenter.stringify_paths(test_file_paths)
+              )
             end
           end
         end
 
-        def self.ensure_spec_opts_have_knapsack_pro_formatters
-          return unless ENV['SPEC_OPTS']
+        attr_reader :cli_args, :rspec_runner
 
-          if [
-              ENV['SPEC_OPTS'].include?(KnapsackPro::Formatters::RSpecQueueSummaryFormatter.to_s),
-              ENV['SPEC_OPTS'].include?(KnapsackPro::Formatters::TimeTracker.to_s),
-          ].all?
-            return
-          end
+        def initialize(adapter_class, args)
+          super(adapter_class)
+          @adapter_class = adapter_class
+          @node_assigned_test_file_paths = []
+          @cli_args = prepare_cli_args(args)
+          @rspec_runner = nil # lazy initialization of RSpec::Core::Runner
+        end
 
-          unless ENV['SPEC_OPTS'].include?(KnapsackPro::Formatters::RSpecQueueSummaryFormatter.to_s)
-            ENV['SPEC_OPTS'] = "#{ENV['SPEC_OPTS']} --format #{KnapsackPro::Formatters::RSpecQueueSummaryFormatter}"
+        # Based on:
+        # https://github.com/rspec/rspec-core/blob/f8c8880dabd8f0544a6f91d8d4c857c1bd8df903/lib/rspec/core/runner.rb#L85
+        #
+        # @return [Fixnum] exit status code.
+        #   0 if all specs passed,
+        #   or the configured failure exit code (1 by default) if specs failed.
+        def run
+          pre_run_setup
+
+          return @rspec_runner.knapsack__exit_early if @rspec_runner.knapsack__wants_to_quit?
+
+          begin
+            exit_code = @rspec_runner.run_specs(self, KnapsackPro.logger)
+          rescue KnapsackPro::Runners::Queue::BaseRunner::TerminationError
+            @rspec_runner.knapsack__error_exit_code
+              .yield_self { Core.error_exit_code(_1) }
+              .yield_self { Kernel.exit(_1) }
```

Comment:
@3v0k4 The `_1` syntax was introduced in Ruby 2.7+

> EDIT: Ruby 2.7+ has changed the syntax from @1 to _1. The article has been updated to reflect this.
> Source: https://medium.com/@baweaver/ruby-2-7-numbered-parameters-3f5c06a55fe4

Ruby 2.7 is already deprecated. https://endoflife.date/ruby

I think we officially should support 3.0+. It is already covered in CI tests.

Reply:
We decided to support Ruby 2.7+ and it's specified in gemspec.